### ISSUE TITLE
 exportFunction() & cloneInto() not available in userscripts

### ISSUE THREAD

**Issue by kekkc** [erosman/support#103]
Hi,

seems  exportFunction() & cloneInto() are not available in a userscript ( https://developer.mozilla.org/en-US/docs/Mozilla/Add-ons/WebExtensions/Sharing_objects_with_page_scripts#exportFunction ).

This seems to be because the userscript requires the extension to make this available for the actual user, which was described by Luca in a bugzilla bugreport:
https://bugzilla.mozilla.org/show_bug.cgi?id=1516356#c1 

Any chance that this will be implemented so that both function can be used in scripts?

**Comment by erosman** [comment]
The userscript API was specially created with [Xray vision](https://developer.mozilla.org/en-US/docs/Mozilla/Tech/Xray_vision) to prevent malware userscripts gaining access to privileged functions via the extension.
The reason is security and FireMonkey uses the userscript API for security and privacy.

Scripts can use any JS function in the page-scope. Why does the content script require `exportFunction()` & `cloneInto()`?

**Comment by kekkc** [comment]
>Why does the content script require exportFunction() & cloneInto()?
Because I need to rewrite some functions, e.g.:
`exportFunction(function (){return true;}, a, {defineAs:'confirm'});`
`exportFunction(function (){return false;}, a, {defineAs:'opener'});`

Reason for that is to prevent that my tabs get "hacked":
https://mathiasbynens.github.io/rel-noopener/

**Comment by erosman** [comment]
Where is your userscript?

**Comment by kekkc** [comment]
Where? It's my own local script, I can PM it to you. But currently doesn't contain much else as a list of these functions that I nullify.

**Comment by erosman** [comment]
Have you tried `window.eval(......)` ?

**Comment by kekkc** [comment]
eval doesn't work for me reliably. 
>exportFunction() & cloneInto() are not JS. They are Firefox internal functions not suitable for content script.

FM is a FF extension ;)

But seriously: your linked GM documention leads to https://wiki.greasespot.net/Content_Script_Injection where it's cleary stated: "Warning: The contents of this page are not accurate when used with Greasemonkey 4.0." 
In GM 4.0 exportFunction() & cloneInto()  are the only way to exchange data with the website, because unsafeWindow was disabled (I guess this includes eval).

>That would be a great security risk if user-script can interact with internal browser functions. 

No, the documentation on https://developer.mozilla.org/en-US/docs/Mozilla/Tech/XPCOM/Language_Bindings/Components.utils.exportFunction describes it:

>This function provides a safe way to expose a function from a privileged scope to a less-privileged scope. In this way privileged code, such as an extension, can share code with less-privileged code like a normal web page script.

**Comment by erosman** [comment]
> FM is a FF extension ;)

Internal API are not content JavaScript. The code designed to run in Firefox background. It is not the same as the JavaScript that runs inside a web page.

For example, none of the `browser.***.***()` API are available inside JS in a web-page.

> But seriously: your linked GM documention leads to https://wiki.greasespot.net/Content_Script_Injection where it's cleary stated: "Warning: The contents of this page are not accurate when used with Greasemonkey 4.0."

True.. that is a bad practice but some people still do it. I did not include the `unsafeWindow` in FireMonkey for that reason. I was showing you alternative methods of communication between userscript & page script.


`In GM 4.0 exportFunction() & cloneInto() are the only way to exchange data with the website, because unsafeWindow was disabled (I guess this includes eval).`

`eval()` is a JavaScript function while `unsafeWindow` is a GreaseMonkey made up function and `exportFunction()` & `cloneInto()` are core Firefox functions designed to run in the browser-scope.

Page/content-scope functions should not be mistaken for browser-scope functions.

FireMonkey does not include unsafe methods that allows malware to cause harm to users by using browser-scope functions (such as `exportFunction()` & `cloneInto()` )

No, the documentation on https://developer.mozilla.org/en-US/docs/Mozilla/Tech/XPCOM/Language_Bindings/Components.utils.exportFunction describes it:

That is used for extension's OWN code which is owned by the extension and checked and has privilege. Extension code is the code that is inside the addon when you download the addon.

It is not to be used for userscript which is not written by the extension developer and not checked by AMO and has unknown code & origin.


userscript is content-scope JavaScript code and can use any function that are part of JavaScript.

userscript

**Comment by kekkc** [comment]
Old thread, but just some addition:
>exportFunction() & cloneInto() are core Firefox functions designed to run in the browser-scope.

No, seems this was not the intention, according to the bugzilla ticket, this use case was considered primarily and should be alot safer & more reliable than e.g. eval:
>The userScripts API already allows the extension itself to provide this kind of helpers (if they are needed and they should be available to the particular user script), because the userScripts's apiScript has these helpers available

**Comment by erosman** [comment]
exportFunction() & cloneInto() are not JavaScript. In order to make them available, background script has to create an API for them specifically.

I have to check with Mozilla security team to see what are the security risks of doing so. If there are no security risks, then it can be considered.

**Comment by erosman** [comment]
Ref: Chat with **Luca Greco** (`userScript` API Developer)

> **erosman:**  I have always tried to make FireMonkey as secure as possible and avoid risky situations. Global window functions like the `unsafeWindow` has been known to be a risk. window.wrappedJSObject` & `exportFunction` & `cloneInto` are similar to `unsafeWindow`. I don't want to open an avenue for exploitation even if that means loosing users.

> **Luca Greco:** yeah, exposing those (`unsafeWindow`, `exportFunction` and `cloneInto`) will make the userScript potentially more exposed to the content, this is definitely true, and so potentially more open to exploitation. We decided not to provide those helpers in the userScripts global by default for the same reasons.

**Comment by erosman** [comment]
I have updated the Help to provide more info on the subject.


Also ..... quoting from above linked "update the userscripts"

> Out of curiosity ........... Can someone ask Jason Barnabe at Greasy Fork, what percentage of scripts require unsafewindow, exportFunction() & cloneInto()?!!!
> If the percentage is high, I will try to work something out with Mozilla security team (somehow).

**Comment by kekkc** [comment]
Thanks for digging into this and consolidating the info here ;) Would be cool if there's an "official" way.

I only found this old blog post that announced ""unsafe-content-script": true" as workaround and exportFunction() & cloneInto() as the "real fix" for userscripts (in 2014 these were content scripts / GM scripts)( https://blog.mozilla.org/addons/2014/04/10/changes-to-unsafewindow-for-the-add-on-sdk/ ). 

Also GM needs a way to turn xrays of, if that is related:
https://bugzilla.mozilla.org/show_bug.cgi?id=1437098#c22

**Comment by erosman** [comment]
Firefox implemented **Xray vision** to make users safer.
Turning off or bypassing **Xray vision** is a risk to the user security and will open up security holes.

`exportFunction()` & `cloneInto()` create security holes. The MDN document explains the issue with reference to extensions own code.

It is even more dangerous if that is allowed for unknown 3rd party scripts (e.g. user-script) as it will create all sorts of problems.

Here are some snippet that I have added to the Help document for the next release.
⬇⬇⬇⬇⬇


  <h3>unsafeWindow</h3>
  <blockquote>Source: <a href="https://sourceforge.net/p/greasemonkey/wiki/unsafeWindow/" target="_blank">unsafeWindow</a><br><br>
    
  This API object allows a User script to access "custom" properties--variable and functions defined in the page--set by the web page. The unsafeWindow object is shorthand for <code>window.wrappedJSObject</code>. It is the raw window object inside the <code>XPCNativeWrapper</code> provided by the Greasemonkey Sandbox.<br><br>
  <b>USE OF UNSAFEWINDOW IS INSECURE, AND IT SHOULD BE AVOIDED WHENEVER POSSIBLE.</b></blockquote>

  <p>Safer alternatives to <code>unsafeWindow</code> are also listed in above page.</p>
  
  <blockquote>Source: <a href="https://wiki.greasespot.net/UnsafeWindow" target="_blank">unsafeWindow</a><br><br>
  
  This command can open certain security holes in your user script, and it is recommended to use this command sparingly.<br>

  Please be sure to read the entire article and understand it before using it in a script.</blockquote>
  
  
  <p>More Info: <a href="https://stackoverflow.com/questions/10824697/why-is-window-and-unsafewindow-not-the-same-from-a-userscript-as-from-a-scrip"  target="_blank">Why is window (and unsafeWindow) not the same from a userscript as from a &lt;script&gt; tag?</a></p>
  

  <h3>Sharing objects with page scripts</h3>  

  <p>Keywords: Xray vision in Firefox, exportFunction(), cloneInto(), Constructors from the page context</p>
  
  <blockquote>Source: <a href="https://developer.mozilla.org/en-US/docs/Mozilla/Add-ons/WebExtensions/Sharing_objects_with_page_scripts#exportFunction" target="_blank">Sharing objects with page scripts</a>

  As an extension developer you should consider that scripts running in arbitrary web pages are hostile code whose aim is to steal the user's personal information, damage their computer, or attack them in some other way.
  
  The isolation between content scripts and scripts loaded by web pages is intended to make it more difficult for hostile web pages to do this.
  
  Since the techniques described in this section break down that isolation, they are inherently dangerous and should be used with great care.</blockquote>

  <h4><a href="https://developer.mozilla.org/en-US/docs/Mozilla/Tech/Xray_vision" target="_blank">Xray Vision</a> & JavaScript Scopes</h4> 
  <ol style="list-style-type: upper-roman;">
    <li> Browser Scope (Trusted privileged code to interact with browser)</li>
    <li> Extension Content Scope (Trusted extension JS injected into web page with API privileges)</li>
    <li> User-Script Scope (Untrusted unverified 3rd party JS injected into web page without direct API privileges)</li>
    <li> Page Scope (JS injected in a web page by the website)</li>
  </ol>

**Comment by erosman** [comment]
Note: 

<a href="https://github.com/Tampermonkey/tampermonkey/blob/07f668cd1cabb2939220045839dec4d95d2db0c8/src/content.js#L519" target="_blank">TemperMonkey <code>unsafeWindow</code></a> Injects code via injecting <code>&lt;script&gt;...&lt;/script&gt;</code>

**Comment by erosman** [comment]
v1,24 Added `unsafeWindow` support

**Comment by kekkc** [comment]
Wow, pretty cool, thx ;)

Realized only that this code produces an error FF 72.02:
`window.wrappedJSObject.focus = function (){};`
"Error: Permission denied to access object"

Strangely this error is not displayed in FF developer 73b12

**Comment by erosman** [comment]
FM `unsafeWindow` also uses `window.wrappedJSObject` (similar to GM).

Let me know how it works out.

**Comment by kekkc** [comment]
Gotch, unsafeWindow works great, no errors. Many many thanks erosman ;)

**Comment by Sxderp** [comment]
I could be mis-remembering (my testing was done so long ago and I haven't touched scripts in a while), but I think `exportFunction` retains line numbers on errors. A lot of things `exportFunction` does can be mimiced with a `window.eval` or `new window.Function('arg1', 'arg2', 'return arg1 + arg2')`, but you lose some debuggability.

Not saying it's worth the security risks (assuming my memory is working properly), just thought I'd point it out.

**Comment by erosman** [comment]
@Sxderp Thanks for the info. I will test it out.

userScript API does log errors to the console with clickable reference to the error location (e.g. script blob), although sometimes errors are generated elsewhere if GM.*/GM_* API is used.

**Comment by Sxderp** [comment]
Here's a sample. Assuming `exportFunction` is available.
```javascript
// ==UserScript==
// @name         meh
// @version      1.0
// @match        https://example.com/*
// ==/UserScript==
function exportFunc() {
  let unknown;
  unknown.error;
}
exportFunction(exportFunc, window, {defineAs: 'exportFunc'});

let funcFunc = new window.Function(`
let unknown;
unknown.error;
`);
window.wrappedJSObject.funcFunc = funcFunc;

window.eval(`
window.evalFunc = function() {
let unknown;
unknown.error;
}
`);
```

Go to https://example.com and open the debug window (Ctrl+Shift+K). In the interpreter tool (I don't know what it's actually called), run `window.exportFunc()`, `window.funcFunc()`, `window.evalFunc()`. The export is the only one with a clickable error line.

**Comment by erosman** [comment]
**Test 1**

```JavaScript
// ==UserScript==
// @name         meh
// @version      1.0
// @match        https://example.com/*
// ==/UserScript==

function exportFunc() {
  let unknown;
  unknown.error;
}

typeof exportFunction !== 'undefined' &&  
       exportFunction(exportFunc, window, {defineAs: 'exportFunc'});

let funcFunc = new window.Function(`
  let unknown;
  unknown.error;
`);
window.wrappedJSObject.funcFunc = funcFunc;

window.eval(`
  window.evalFunc = function() {
  let unknown;
  unknown.error;
  }
`);

console.log('window.exportFunc', typeof window.exportFunc);
console.log('window.funcFunc', typeof window.funcFunc);
console.log('window.evalFunc', typeof window.evalFunc);
```

![image](https://user-images.githubusercontent.com/6458876/74079882-89545300-4a52-11ea-85c0-4ef407c0eb2b.png)


All above console.log are clickable to the blob.

**Comment by erosman** [comment]
Note: code errors also have clickable links

![image](https://user-images.githubusercontent.com/6458876/74079993-21067100-4a54-11ea-863f-e7161a3a052d.png)

**Comment by erosman** [comment]
**Test 2**

```JavaScript
// ==UserScript==
// @name         meh
// @version      1.0
// @match        https://example.com/*
// ==/UserScript==

function exportFunc() { console.log('this is exportFunc'); }

typeof exportFunction !== 'undefined' &&  
    exportFunction(exportFunc, window, {defineAs: 'exportFunc'});
console.log('exportFunc', typeof window.exportFunc);

let funcFunc = new window.Function(`console.log('this is funcFunc');`);
console.log('funcFunc', typeof window.exportFunc);

window.wrappedJSObject.funcFunc = funcFunc;
console.log('funcFunc', typeof window.exportFunc);

window.eval(`function evalFunc() { console.log('this is evalFunc'); }`);
console.log('window.evalFunc', typeof window.evalFunc);

```

![image](https://user-images.githubusercontent.com/6458876/74080011-7a6ea000-4a54-11ea-96bd-6dc47a02107e.png)


AFA how to actually export function without `exportFunction()` I have to do more testing.

**Comment by erosman** [comment]
**Test 3**

```JavaScript
// --- Page JavaScript
var foo = "I'm defined in a page script!";
function runTest() {
  console.log('runTest', foo); // runTest new unsafeWindow foo
  console.log('testFunc', typeof window.testFunc); //  testFunc function
  window.testFunc(); // Error: Permission denied to access object
 testFunc(); // Error: Permission denied to access object
}
```

```JavaScript
// --- user-script
console.log('unsafeWindow.foo', unsafeWindow.foo); // unsafeWindow.foo I'm defined in a page script!
unsafeWindow.foo = 'new unsafeWindow foo'; // changing foo on page
unsafeWindow.testFunc = function() { console.log('this is testFunc'); } // creating new function
unsafeWindow.runTest(); 
unsafeWindow.testFunc(); // this is testFunc
testFunc(); // Error: Permission denied to access object
```

It seems...

- `unsafeWindow` can be used to create functions and export to page window
- page window won't have access to `unsafeWindow` created functions
- `unsafeWindow` cna take the place of `exportFunction()`

re: `cloneInto()` I will test object creation
Note: All data relates to FireMonkey only.

**Comment by erosman** [comment]
**Test 4**

Accessing Page DOM

```JavaScript
unsafeWindow.testFunc = function() { console.log(location.href); }
unsafeWindow.testFunc(); // https://example.com/
```

**Comment by erosman** [comment]
**Test 5**

Normal `cloneInto()`
```JavaScript
const obj = {text: 'hello from user-script'};
unsafeWindow.obj = cloneInto(obj, unsafeWindow);
```

```JavaScript
unsafeWindow.obj = {text: 'hello from user-script'};
console.log(unsafeWindow.obj.text); // hello from user-script
```

**Comment by Sxderp** [comment]
After an eval you must use `window.wrappedJSObject` (assuming this is `unsafeWindow` in FM) to access the object in the UserScript.

```javascript
window.eval(`
window.evalFunc = function() {
let unknown;
unknown.error;
}
`);
console.log(window.wrappedJSObject.evalFunc); // function evalFunc()
```

My concern is that you can't debug the content that was evaled. More complete example.
Page code
```javascript
function runTestTest() {
  console.log('runTestEval');
  console.log('evalFunc', typeof window.evalFunc);
  window.evalFunc(); // This errors with no clickable bit
}
```

Userscript
```javascript
// ==UserScript==
//
// ==/UserScript==
window.eval(`
window.evalFunc = function() {
console.log('in eval func')
let unknown;
unknown.error;
console.log('out eval func')
}
`);

window.wrappedJSObject.runTestEval();
```

No error message appears in the web page console. An error does appear in the browser console, but it is not clickable.

Now, this is generally pretty edge case. I'd say it only occurs when you're purposely piggy-backing off of the page content, or trying to overwrite very specific functions within the page content.

**Comment by erosman** [comment]
I actually get a clickable reference however, since the actual function in on the page, it refers to the page (not the user-script that called it)

![image](https://user-images.githubusercontent.com/6458876/74090425-75dfd100-4ac0-11ea-8168-a58f6f693273.png)

**Comment by Sxderp** [comment]
That's weird, how do you get evalFunc is not a function? Using Firemonkey with the above code blocks this is what I get in the page console; the eval link (console.log) is clickable, and sends you to the blob. But as you can see the page doesn't indicate any errors.

![Screenshot from 2020-02-08 13-47-33](https://user-images.githubusercontent.com/6609453/74090481-d3046400-4a79-11ea-8fd4-fcb1c59b9f82.png)

And in the browser console; the eval link is not clickable.

![Screenshot from 2020-02-08 13-48-37](https://user-images.githubusercontent.com/6609453/74090483-d7c91800-4a79-11ea-87d5-aafb29832636.png)

**Comment by Sxderp** [comment]
> That creates a no-reference error
> 

It's supposed to. In order for force an error.

**Comment by erosman** [comment]
Page Script

```JavaScript
function runTestEval() {
  console.log('runTestEval');
  console.log('evalFunc', typeof window.evalFunc);
  window.evalFunc(); // This errors with no clickable bit
}
```

User script (I commented out the error)

```JavaScript
window.eval(`
window.evalFunc = function() {
console.log('in eval func')
let unknown;
//unknown.error;
console.log('out eval func')
}
`);

window.wrappedJSObject.runTestEval();
```
All are clickable to the page or debugger

![Clipboard01](https://user-images.githubusercontent.com/6458876/74090566-14b8fd00-4ac2-11ea-98be-40593514279b.jpg)

**Comment by Sxderp** [comment]
> All are clickable to the page or debugger
> 

Yes. My concern was that *errors* are not clickable (nor do they show up in the page console). I would not call it a major issue, but it does make debugging a little harder.

**Comment by erosman** [comment]
> That's weird, how do you get evalFunc is not a function?

Sorry... That was me. :p
I was testing and didnt define it to see how the error comes up.

> My concern was that errors are not clickable (nor do they show up in the page console)

Sadly, it is out of our hands ;)

**Comment by kekkc** [comment]
Sry to bother (again), but seems there's still some glitch, if the following userscript is used (all frames have to be used):
`// ==UserScript==
// @name 		asdf
// @run-at 		document-start
// @match 		<all_urls>
// @allFrames	true
// ==/UserScript==
unsafeWindow.document.hasFocus = function() {return true;};`
on the following site:
https://www.w3schools.com/jsref/tryit.asp?filename=tryjsref_document_hasfocus
webconsole gets flooded with "Error: Permission denied to access object" that is called repeatedly & endlessly. In general, I guess there shouldn't be any issue, since the script should be inserted in the frame and therefore same-origin.

**Comment by Sxderp** [comment]
Try:
```javascript
// ==UserScript==
// ...
// ==/UserScript==
let hasFocus = new window.Function('return true');
unsafeWindow.document.hasFocus = hasFocus;
```

**Comment by erosman** [comment]
@kekkc To start with, that is a really poor code on `w3schools.com`
```JavaScript
setInterval("myFunction()", 1);

function myFunction() {
  var x = document.getElementById("demo");
  if (document.hasFocus()) {
    x.innerHTML = "The document has focus.";
  } else {
    x.innerHTML = "The document DOES NOT have focus.";
  }
}
```

That `setInterval()` is running 1000 times per second. Who wrote that!?? :-1: 

Then the error comes from that page (not the user-script)

```
Error: Permission denied to access object               tryit.asp:14:16
->  if (document.hasFocus())
```

I tested it with GM and the same error come up, although the error only comes up when clicking as it breaks codemirror code.

```
Error: Permission denied to access object 5               codemirror.js:3702:90
-> var takeFocus = op.focus && op.focus == activeElt() && (!document.hasFocus || document.hasFocus())
```

**Comment by kekkc** [comment]
> 
> 
> Try:
> 
> ```js
> // ==UserScript==
> // ...
> // ==/UserScript==
> let hasFocus = new window.Function('return true');
> unsafeWindow.document.hasFocus = hasFocus;
> ```

Many thanks, that fixed it. BTW: I also included:
```Object.defineProperty(unsafeWindow.document, 'hidden', {value: false});```

Maybe these methods of "overwriting" methods can also be included in the help file ;)

>I tested it with GM and the same error come up, although the error only comes up when clicking as it breaks codemirror code.

Thanks, I realized the same. Know that w3schools.com is often bad, but at least good for tests. 

_OT: Nevertheless Sxderp's suggestion fixed the "Error: Permission denied to access object" (and many other issues I had with my old method of replacing methods. To be exact 20 years old, coming from an old program [Proxomitron](http://www.buerschgens.de/Prox/) that focused on filtering / modifying webpages to get rid of ActiveX phone dialers)_
